bftengine/src/bftengine/MessageBase.cpp

```diff
@@ -12,183 +12,230 @@
+  Assert(owner_);
+  Assert(msgSize_ > 0);
 
-		size_t MessageBase::sizeNeededForObjAndMsgInLocalBuffer() const
-		{
-			Assert(owner_);
-			Assert(msgSize_ > 0);
+  const size_t sizeNeeded = sizeof(RawHeaderOfObjAndMsg) + msgSize_;
 
-			const size_t sizeNeeded = sizeof(RawHeaderOfObjAndMsg) + msgSize_;
+  return sizeNeeded;
+}
 
-			return sizeNeeded;
-		}
+MessageBase *MessageBase::createObjAndMsgFromLocalBuffer(char *buffer,
+                                                         size_t bufferLength,
+                                                         size_t *actualSize) {
+  if (actualSize) *actualSize = 0;
 
-		MessageBase* MessageBase::createObjAndMsgFromLocalBuffer(char* buffer, size_t bufferLength, size_t* actualSize)
-		{
-			if(actualSize) *actualSize = 0;
+  if (bufferLength <= sizeof(RawHeaderOfObjAndMsg)) return nullptr;
 
-			if (bufferLength <= sizeof(RawHeaderOfObjAndMsg)) return nullptr;
+  RawHeaderOfObjAndMsg *pHeader = (RawHeaderOfObjAndMsg *) buffer;
+  if (pHeader->magicNum != magicNumOfRawFormat) return nullptr;
+  if (pHeader->msgSize == 0) return nullptr;
+  if (pHeader->msgSize > maxExternalMessageSize) return nullptr;
+  if (pHeader->msgSize + sizeof(RawHeaderOfObjAndMsg) > bufferLength)
+    return nullptr;
 
-			RawHeaderOfObjAndMsg* pHeader = (RawHeaderOfObjAndMsg*)buffer;
-			if (pHeader->magicNum != magicNumOfRawFormat) return nullptr;
-			if (pHeader->msgSize == 0) return nullptr;
-			if (pHeader->msgSize > maxExternalMessageSize) return nullptr;
-			if (pHeader->msgSize + sizeof(RawHeaderOfObjAndMsg) > bufferLength) return nullptr;
-			
-			char* pBodyInBuffer = buffer + sizeof(RawHeaderOfObjAndMsg);
+  char *pBodyInBuffer = buffer + sizeof(RawHeaderOfObjAndMsg);
 
-			void* msgBody = std::malloc(pHeader->msgSize);
-			memcpy(msgBody, pBodyInBuffer, pHeader->msgSize);
+  void *msgBody = std::malloc(pHeader->msgSize);
+  memcpy(msgBody, pBodyInBuffer, pHeader->msgSize);
 
-			MessageBase* msgObj =
-				new MessageBase(pHeader->sender, (MessageBase::Header*)msgBody, pHeader->msgSize, true);
+  MessageBase *msgObj =
+      new MessageBase(pHeader->sender,
+                      (MessageBase::Header *) msgBody,
+                      pHeader->msgSize,
+                      true);
 
-			if (actualSize) *actualSize = (pHeader->msgSize + sizeof(RawHeaderOfObjAndMsg));
+  if (actualSize)
+    *actualSize = (pHeader->msgSize + sizeof(RawHeaderOfObjAndMsg));
 
-			return msgObj;
-		}
+  return msgObj;
+}
 
+bool MessageBase::equals(const MessageBase &other) const {
+  bool equals = (other.msgSize_ == msgSize_ && other.storageSize_ == storageSize_ &&
+      other.sender_ == sender_ && other.owner_ == owner_);
+  if (!equals)
+    return false;
+  if (!other.msgBody_ && !msgBody_) // Both are nullptr => OK
+    return true;
+  if (!other.msgBody_ || !msgBody_) // Only one is nullptr => NOK
+    return false;
+  return (*other.msgBody_ == *msgBody_);
```

Comment:
Only the headers will be compared. I assume you want to compare the whole message

Reply:
You are right. In case msgBody_ points to the whole message including the header, all allocated memory should be compared and bool operator==(const Header &other) const is probably not required.